extensions/message-tags.md

```diff
@@ -1,97 +1,143 @@
 
-The final version of the specification will use an unprefixed capability name.
+## Format
 
-## Introduction
+The message pseudo-BNF, as defined in [RFC 1459, section 2.3.1][rfc1459] is extended as follows:
 
-This specification extends the [IRCv3.2 Message Tags](./message-tags-3.2.html) specification in a number of ways:
+    <message>       ::= ['@' <tags> <SPACE>] [':' <prefix> <SPACE> ] <command> <params> <crlf>
+    <tags>          ::= <tag> [';' <tag>]*
+    <tag>           ::= <key> ['=' <escaped value>]
+    <key>           ::= [ '+' ] [ <vendor> '/' ] <sequence of letters, digits, hyphens ('-')>
+    <escaped value> ::= <sequence of any characters except NUL, CR, LF, semicolon (`;`) and SPACE>
+    <vendor>        ::= <host>
 
-* Adds a new capability to indicate base support for message tags
-* Defines a prefix for expressing client-only tags
-* Defines a new command for tag-only messages
-* Increases the size limit for tags
+Not all tag keys have a value.
 
-## Motivation
+The ordering of tags is not meaningful.
 
-Previously, clients were required to negotiate a capability with servers for each
-supported tag. This made tags inappropriate for client-only features. By adding a
-new base capability, this specification allows clients to indicate support for
-receiving any well-formed tag, whether or not it is recognised or used. This also
-frees servers from having to filter individual message tags for each client
-response.
+Individual tag keys MUST only be used a maximum of once per message. Clients
+receiving messages with more than one occurrence of a tag key SHOULD discard all
+but the final occurrence.
 
-The client-only tag prefix allows servers to safely relay untrusted client tags,
-keeping them distinct from server-initiated tags that carry verified meaning.
+## Escaping values
+
+The mapping between characters in tag values and their representation in `<escaped value>` is defined as follows:
+
+| Character       | Sequence in `<escaped value>` |
+|-----------------|-------------------------------|
+| `;` (semicolon) | `\:` (backslash and colon)    |
+| `SPACE`         | `\s`                          |
+| `\`             | `\\`                          |
+| `CR`            | `\r`                          |
+| `LF`            | `\n`                          |
+| all others      | the character itself          |
+
+This escape format is more space efficient than URL-escaping, and ensures that message parts can easily by split on spaces and semi-colons before further parsing.
+
+If a lone `\` exists at the end of an escaped value (with no escape character following it), then there
+SHOULD be no output character. For example, the escaped value `test\` should unescape to `test`.
+
+## Size limit
+
+The size limit for message tags is 4607 bytes, including the leading `'@'` (0x40) and trailing space `' '` (0x20) characters. The size limit for the rest of the message remains unchanged at 512 bytes.
```

Comment:
For tags that aren't using the additional CAP, you'll never reach the 512 limit, so it doesn't make much practical difference. You won't ever be sent more than that if you don't negotiate it.

Reply:
The *only* issue I can see here is current implementations maybe worrying they might not be “compliant” if they do any sort of limit checking as a hard coded 512. But practically it won’t ever be an issue for them.

Also, the new wording covers them fine: server tags are still limited to 512, and that’s all they’ll ever see.

Reply:
Hmm, yeah we are gonna have those grandfathered impls that don't comply with the spec exactly as written (_"Clients and servers that negotiate a capability that uses tags MUST support the full tag specification"_ they don't support this size limit part as written). Not gonna be an issue for them, but still worth keeping in mind as we push this forwards. I don't have a big drama with the line above or requiring it for future tag capabilities though, feels like it should be fine, would appreciate input from other server implementers on the change.

Reply:
Again, it’s not a change :)

Reply:
It is a change from the [stock IRCv3.2 message tags spec](https://ircv3.net/specs/core/message-tags-3.2.html) used by account-tag and server-time. It's not an impactful change or one that will break running servers but it is a change. With just that spec, tag space by the spec is `512`. With this new spec, tag space by the spec is `4094+512`. It doesn't actively impact anything since the servers implementing just those tags aren't gonna hit the limit, but they will not comply 100% with the spec (and it's important to keep that in mind so we don't run into something like the Metadata spec and get any backlash around shifting compatibility/requirements).

Reply:
It’s a zen change. If a tree falls in the wood with no one around to hear it type thing. Not worth worrying about or worrying about other people worrying about :)

Reply:
Sure, just keep it in mind